### PR TITLE
feat: Modified Dynamic Fee Handlers with twap oracle

### PR DESCRIPTION
**Pull request by andersonlee725**
Modified Dynamic Fee Handlers calculate the fee in the native coin based on the twap oracle.

## Description
Created a twap oracle using Uniswap V3.
In the Dynamic Fee Handlers, calculate the tx cost fee of the destination chain and the fee is converted to the amount in the native coin of the resource chain.
For converting, the twap oracle is used.

## Related Issue Or Context

Closes: #<issue>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [ ] I have updated the documentation locally and in chainbridge-docs.
- [ ] I have added tests to cover my changes.
- [ ] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

**Review by viatrix — CHANGES_REQUESTED** [review]

**Review by viatrix — APPROVED** [review]

**Review by viatrix — APPROVED** [review]

File: contracts/handlers/fee/V2/DynamicFeeHandlerV2.sol [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    function renounceAdmin(address newAdmin) external {
+        address sender = _msgSender();
+        require(sender != newAdmin, 'Cannot renounce oneself');
+        grantRole(DEFAULT_ADMIN_ROLE, newAdmin);
+        renounceRole(DEFAULT_ADMIN_ROLE, sender);
+    }
+
+    /**
+        @notice Sets the fee oracle address for signature verification.
```
Comment (lastperson) [inline contracts/handlers/fee/V2/DynamicFeeHandlerV2.sol:78]: Outdated notice.

**Review by lastperson — CHANGES_REQUESTED** [review]
Need to account for possibility where there is no destination domain wrap pool available. In such case it should be possible to manually set the price rate.

File: contracts/handlers/fee/V2/TwapOracle.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Factory.sol';
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import "../../../utils/TickMath.sol";
+import "../../../utils/FullMath.sol";
+import "../../../utils/AccessControl.sol";
+
+contract TwapOracle is AccessControl {
+    IUniswapV3Factory public immutable UNISWAP_V3_FACTORY;
+    address public immutable WETH;
+    uint24[] internal _knownFeeTiers;
```
[the remark this answers is not in the document]
Comment (andersonlee725) [reply contracts/handlers/fee/V2/TwapOracle.sol:14]: Will remove it.

File: contracts/utils/TickMath.sol [hunk 3]
```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+/// @title Math library for computing sqrt prices from ticks and vice versa
```
[the remark this answers is not in the document]
Comment (andersonlee725) [reply contracts/utils/TickMath.sol:4]: The lib from the default dependency uses solidity >=0.5.0 <0.8.0.
So I had to copy from another branch of openzeppelin.

File: contracts/handlers/fee/V2/TwapOracle.sol [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    uint32 internal _timeWindow;
+    mapping(address => mapping(address => address)) public pools;
```
[the remark this answers is not in the document]
Comment (andersonlee725) [reply contracts/handlers/fee/V2/TwapOracle.sol:17]: Okay, will make it to single mapping and add the time window into the mapping as well.

**Review by viatrix — APPROVED** [review]

**Review by lastperson — APPROVED** [review]

**Review by nmlinaric — APPROVED** [review]

**Pull request merged**